sdk/dynamicconfig/sdk/metric/sdk.go

```diff
@@ -0,0 +1,625 @@
+			// strategy to wait until Collect() removes this entry from the map.
+			//
+			// This can be improved by having a list of "Unmapped" entries for
+			// one time only usages, OR we can make this a blocking path and use
+			// a Mutex that protects the delete operation (delete only if the old
+			// record is associated with the key).
+
+			// Let collector get work done to remove the entry from the map.
+			runtime.Gosched()
+			continue
+		}
+		// The new entry was added to the map, good to go.
+		return rec
+	}
+}
+
+func (s *syncInstrument) Bind(kvs []kv.KeyValue) api.BoundSyncImpl {
+	return s.acquireHandle(kvs, nil)
+}
+
+func (s *syncInstrument) RecordOne(ctx context.Context, number api.Number, kvs []kv.KeyValue) {
+	h := s.acquireHandle(kvs, nil)
+	defer h.Unbind()
+	h.RecordOne(ctx, number)
+}
+
+// NewAccumulator constructs a new Accumulator for the given
+// processor.  This Accumulator supports only a single processor.
+//
+// The Accumulator does not start any background process to collect itself
+// periodically, this responsbility lies with the processor, typically,
+// depending on the type of export.  For example, a pull-based
+// processor will call Collect() when it receives a request to scrape
+// current metric values.  A push-based processor should configure its
+// own periodic collection.
+func NewAccumulator(processor export.Processor, opts ...Option) *Accumulator {
+	c := &Config{}
+	for _, opt := range opts {
+		opt.Apply(c)
+	}
+
+	return &Accumulator{
+		processor:        processor,
+		asyncInstruments: internal.NewAsyncInstrumentState(),
+		resource:         c.Resource,
+		ext: c.Ext,
+	}
+}
+
+// NewSyncInstrument implements api.MetricImpl.
+func (m *Accumulator) NewSyncInstrument(descriptor api.Descriptor) (api.SyncImpl, error) {
+	return &syncInstrument{
+		instrument: instrument{
+			descriptor: descriptor,
+			meter:      m,
+		},
+	}, nil
+}
+
+// NewAsyncInstrument implements api.MetricImpl.
+func (m *Accumulator) NewAsyncInstrument(descriptor api.Descriptor, runner metric.AsyncRunner) (api.AsyncImpl, error) {
+	a := &asyncInstrument{
+		instrument: instrument{
+			descriptor: descriptor,
+			meter:      m,
+		},
+	}
+	m.asyncLock.Lock()
+	defer m.asyncLock.Unlock()
+	m.asyncInstruments.Register(a, runner)
+	return a, nil
+}
+
+// Collect traverses the list of active records and observers. If the periods
+// option is set with more than one period, it only exports metrics with a
+// CollectionPeriod that matches one of the periods. If not, it exports data
+// for each active instrument.
+//
+// Collect() may not be called concurrently.
+//
+// During the collection pass, the export.Processor will receive
+// one Export() call per current aggregation.
+//
+// Returns the number of records that were checkpointed.
+func (m *Accumulator) Collect(ctx context.Context, opts ...CollectOption) int {
+	c := &CollectConfig{}
+	for _, opt := range opts {
+		opt.Apply(c)
+	}
```

Comment:
It seems that the only option being applied is the set of collection periods. Would this likely change between calls to Collect? If not, would it be possible to move this logic to the New method, saving some overhead in reapplying the same configs?

Reply:
Explained here: https://github.com/vmingchen/opentelemetry-go-contrib/pull/2#discussion_r453697881

So the option would likely be different for every call of Collect. I renamed `CollectConfig` to `CollectOptions` instead. I structured things this way, because I don't want to change the Collect interface, and break old code, so I needed `periods` to be an optional parameter.